src/lib/sql/sql_pipeline.cpp

```diff
@@ -0,0 +1,136 @@
+#include <utility>
+
+#include "SQLParser.h"
+#include "concurrency/transaction_manager.hpp"
+#include "logical_query_plan/lqp_translator.hpp"
+#include "optimizer/optimizer.hpp"
+#include "sql_pipeline.hpp"
+#include "sql_translator.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+SQLPipeline::SQLPipeline(const std::string& sql, std::shared_ptr<TransactionContext> transaction_context) : _sql(sql) {
+  if (transaction_context != nullptr) {
+    _transaction_context = transaction_context;
+  } else {
+    _transaction_context = TransactionManager::get().new_transaction_context();
+  }
+}
+
+const hsql::SQLParserResult& SQLPipeline::get_parsed_sql() {
+  if (_parsed_sql) {
+    // Returned cached result
+    return *(_parsed_sql.get());
+  }
+
+  auto parse_result = std::make_unique<hsql::SQLParserResult>();
+  try {
+    hsql::SQLParser::parse(_sql, parse_result.get());
+  } catch (const std::exception& exception) {
+    throw std::runtime_error("Error while parsing SQL query:\n  " + std::string(exception.what()));
+  }
+
+  if (!parse_result->isValid()) {
+    throw std::logic_error("SQL query not valid.");
+  }
+
+  _parsed_sql = std::move(parse_result);
+  return *(_parsed_sql.get());
+}
+
+const std::vector<std::shared_ptr<AbstractLQPNode>>& SQLPipeline::get_unoptimized_logical_plan(const bool validate) {
+  if (!_unopt_logical_plan.empty()) {
+    return _unopt_logical_plan;
+  }
```

Comment:
If I call `get_unoptimized_logical_plan` twice, using different values for `validate`, this breaks.